Instance norm (and batch, layer norm) using NVFuser Python frontend only

jacobhinkle opened the pull request:
This implements a generic norm layer using the NVFuser Python frontend without using the CUDAExtension introduced in the target branch. I translated the pieces that were not exposed in `fd.ops`, namely the `instance_norm` and `instance_norm_backward` functions in `ops/normalization.cpp`.

The code is modified a bit to take a `stat_axes` parameter, which indicates the axes which are used to index the computed statistics. For instance norm, the stats are computed for each combination of instance (batch dim) and channel, so `stat_axes = [NamedAxis.BATCH, NamedAxis.CHANNEL]`. Batch and layer norm correspond to only the channel or batch dimension, respectively. The batch dimension is always reduced when computing running stats, but not the channel dim. I have not tested batch or layer norm using this approach yet as I have so far avoided modifying the test in the target branch at all.

Note that this PR requires using the `devel` branch of PyTorch (for `add_output(t, alias_input=s)`) which has undergone refactoring of nvfuser into `thirdparty/`.

eqy committed:
initial check in

eqy committed:
add weight and bias check

eqy committed:
address comments, cleanup

eqy committed:
fix

eqy committed:
sketchy test numerics twiddling

eqy committed:
add profile, remove scalars from cache key

eqy committed:
retab

eqy committed:
some overdue cleanup

eqy committed:
fix device for dummy tensor

eqy committed:
add test for multigpu instancenorm3dnvfuser

jpool-nv committed:
[contrib][sparsity] Conv1 d support (#1466)

* support Conv1D layers in both pruning and permutation steps

* remove temporary prints

Fuzzkatt committed:
update fused_dense to use torch.autocast instead of apex amp (#1490)

* update fused_dense to use torch.autocast instead of apex amp

* use _cast_if_autocast_enabled for inputs

* remove extraneous line

* update test_fused_dense, remove assert_allclose -> assert_close, test float32 as well

* remove extraneous cast

kumpera committed:
Replace usage of internal with public APIs (#1496)

* Update distributed_fused_adam.py to use public function

PT master now has get_global_rank as part of the public API, switch to it if available

* Update distributed_fused_lamb.py to use public function

crcrpar committed:
Update for CUDA 11.8 (#1502)

* Correct handling of nvcc version to SM version map

Fixes sm_90 building for extensions that don't obey `TORCH_CUDA_ARCH_LIST`.

* Fix fMHA to allow sm_90

Co-authored-by: Vasudevan Rengasamy <vrengasamy@nvidia.com>

* add packaging

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: Cliff Woolley <cliffwoolley@users.noreply.github.com>
Co-authored-by: Vasudevan Rengasamy <vrengasamy@nvidia.com>

crcrpar committed:
check `CUDA_HOME` before using it (#1499)

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Deprecate outdated features (#1506)

* add warnings

update README

verify deprecated warning

deprecate apex.RNN which is never a thing

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
[transformer] Allow for skipping stream sync (#1505)

* Optionally disable stream synchronization after batched p2p communication

* Add test cases with `sync_batch_comm=False`

only when https://github.com/pytorch/pytorch/pull/82450 is included in
pytorch.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* utilize existing test methods

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* consistent naming

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: Aidyn-A <Aidyn-A@users.noreply.github.com>

* silly boy, to skip the sync, set False

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* Test with async_pipelinign w/o sync after batch_isend_irecv

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* again, set sync_batch_comm to False

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: Aidyn-A <Aidyn-A@users.noreply.github.com>

* Remove `torch.testing._internal.common_cuda`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: Sangkug Lym <slym@nvidia.com>
Co-authored-by: Aidyn-A <Aidyn-A@users.noreply.github.com>

Aidyn-A committed:
add tearDown (#1508)

timmoon10 committed:
Fix bug with grad clipping and distributed Adam (#1512)

* Fix bug in dist Adam when grad scaling and grad clipping are applied together

Grads must be unscaled before grad clipping.

* Use no-copy NCCL collectives in distributed Adam

Avoid allocating temporary buffers in collectives.

* Avoid excessive reduce-scatter synchronizations in distributed Adam

* Support dist Adam optimizer in Megatron grad scaler

Review suggestion from @crcrpar.

crcrpar committed:
Partial activation checkpoint (#1511)

Co-authored-by: Sangkug Lym <slym@nvidia.com>

Aidyn-A committed:
Use all_gather_into_tensor and reduce_scatter_tensor (#1513)

* update distributed_fused_lamb

* update distributed_fused_lamb

* add test

* update apex transformer

* update test_distributed_fused_lamb

* update test_distributed_fused_lamb

* update test_distributed_fused_lamb

* apply suggested changes

timmoon10 committed:
Support overlapped grad sync with Megatron interleaved pipeline parallelism (#1514)

* Add option to use no_sync context with interleaved pipeline parallelism

* Add unit test for no_sync context with interleaved pipeline parallelism

* Debug no_sync context support in interleaved pipeline parallelism

Fuzzkatt committed:
update run_transformer tests to default to using pytorch native UCC instead of torch_ucc (#1495)

* update HAS_TORCH_UCC to TORCH_UCC

* add comments for failing tests

* move HAS_UCC to _ucc_utils.py

* whitespace

* small changes

* newline

* updated list of failing tests

* update failing tests list

JimmyZhang12 committed:
Remove weight gradients if model is frozen for linear layer (#1517)

* remove wgrad if weights do not require grad

* add seq parallelism support for frozen linear layer

* Update apex/transformer/tensor_parallel/layers.py

Co-authored-by: Masaki Kozuki <masaki.kozuki.2014@gmail.com>

* Update apex/transformer/tensor_parallel/layers.py

Co-authored-by: Masaki Kozuki <masaki.kozuki.2014@gmail.com>

* change for #1513

Signed-off-by: Jimmy Zhang <jiemingz@nvidia.com>

Signed-off-by: Jimmy Zhang <jiemingz@nvidia.com>
Co-authored-by: Jimmy Zhang <jiemingz@nvidia.com>
Co-authored-by: Masaki Kozuki <masaki.kozuki.2014@gmail.com>

crcrpar committed:
check `is_ucc_available` (#1523)


Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
update exception message (#1524)

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
[README] Add a list of install options (#1492)

* add a list of install options

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

* verbose about `fused_layer_norm_cuda` and `fast_layer_norm`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

timmoon10 committed:
Perform dist Adam communication in side stream (#1528)

Avoid bug where main stream is blocked by async NCCL collectives. Stop batching async grad reductions. Defer synchronizing async grad reductions as long as possible if using a contiguous grad buffer.

timmoon10 committed:
add setter for virtual world size (#1541)

Signed-off-by: ericharper <complex451@gmail.com>

Signed-off-by: ericharper <complex451@gmail.com>
Co-authored-by: ericharper <complex451@gmail.com>

timmoon10 committed:
Optimize peer memory halo exchange kernel

Use an alternating double-buffer scheme to allow push-only communication. Remove cooperative group syncs and memory fences.

crcrpar committed:
[contrib] fix `reduce_scatter_tensor` fallouts in DistributedFusedLamb (#1536)

* Add RNN-T test case to dist LAMB

* let tests pass

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: Mikołaj Błaż <mblaz@nvidia.com>

timmoon10 committed:
Avoid index bookkeeping in halo exchange kernel if data is contiguous

yaoyu-33 committed:
Update megatron fused softmax follow megatron-lm (#1539)

* Update megatron fused softmax follow megatron-lm

Signed-off-by: Yu Yao <yuya@nvidia.com>

* Add mask=None support in scaled_masked_softmax

Signed-off-by: Yu Yao <yuya@nvidia.com>

* Update setup.py for scaled_softmax_cuda

Signed-off-by: Yu Yao <yuya@nvidia.com>

* Add tests for fused_scale_softmax (mask=None)

Signed-off-by: Yu Yao <yuya@nvidia.com>

* Assert grad equal in fused softmax test

Signed-off-by: Yu Yao <yuya@nvidia.com>

* Revert "Assert grad equal in fused softmax test"

Signed-off-by: Yu Yao <yuya@nvidia.com>

Signed-off-by: Yu Yao <yuya@nvidia.com>
Co-authored-by: Yu Yao <yuya@nvidia.com>

thorjohnsen committed:
Merge pull request #1543 from timmoon10/peer-halo-exchange-perf-optimization

LGTM

Fuzzkatt committed:
Refactor run_transformer bert minimal and gpt minimal tests (#1540)

* working test_bert_minimal.py

* remove some debugging statements

* working test_gpt_minimal.py

* test_dynamic_batchsize.py having issues with torch.backends.cudnn.allow_tf32

* working test_dynamic_batchsize.py

* refactor test_bert_minimal.py, need to investigate rng of MANUAL_SEED for nccl only pipeline with virtual_pipeline_model_parallel_size = 2

* add test_bert_minimal_alt.py for visibility

* update test_gpt_minimal.py

* lint

* update loss cutoff for bert test

* split with / without interleaving tests for bert

* use skipTest

* remove ONCE

* add ignore_unknown_args=True

* remove old testing files

* add num_devices logic to override_args

crcrpar committed:
change param of list of tensor to tensor for _make_nccl_premul_sum (#1542)

ref:
https://github.com/pytorch/pytorch/commit/63e16216d8830b6340816c873b035e1a31ad4636

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

eqy committed:
Update instance_norm.py

jacobhinkle committed:
Convert instance norm to use Python interface

This removes the need for the C++ PyTorch extension.

jacobhinkle committed:
Generalize instance_norm forward to norm_fusion_forward

This should be a general norm layer forward fusion definition that
covers batch, layer, and instance normalization. Next I will factor out
the backward in a similar manner. Note that only instance normalization
is tested at this point using this general function.

jacobhinkle committed:
Generalize _InstanceNormNVFuser to _NormNVFuserBase

I have not tested the batch or layer norm implementations using this
method. We should have those tests in place as well as fixing the half
precision tests.

jacobhinkle committed:
Remove stale __all__ in instance_norm.py

jacobhinkle commented:
Note this branch is broken as is, since the compile will fail due to headers being in different places if you are using `devel`. I can push the changes that make things compile against devel (it's just a few lines), but thought that should maybe be a separate PR for adapting to the refactor once it's merged upstream.

crcrpar commented on apex/normalization/instance_norm.py:
```diff
@@ -1,83 +1,594 @@
+            inputs = [x]
+            if weight is not None:
+                assert bias is not None
+                tv_weight = fd.define_tensor(weight.ndim, torch2datatype(weight.dtype))
+                tv_bias = fd.define_tensor(bias.ndim, torch2datatype(bias.dtype))
```
How will this behave when `weight` and `bias` are `None`?

jacobhinkle replied:
In that case, both `tv_weight` and `tv_bias` will be set to `None`. I am not using dummies in this implementation, so they are Python `None`s. I should probably handle `weight` and `bias` separately since someone could just use one or the other though...

jacobhinkle committed:
Remove previous C++ instancenorm implementation

jacobhinkle committed:
Cache Fusion objects in NormNVFuserFunction fwd/bwd

jacobhinkle committed:
Fix type hints in instance_norm.py

Mypy no longer complains. However, type hints are still lacking due to
lack of py.typed stubs in PyTorch.

yaoyu-33 committed:
Extend softmax fusion seq length to 16384 (#1558)

Signed-off-by: yaoyu-33 <yaoyu.094@gmail.com>

timmoon10 committed:
Optimizing distributed Adam when running with one work queue (#1551)

* Coalesce reduce-scatters in distributed Adam

* Support variable-size param buckets in dist Adam optimizer

* Support contiguous grad buffer with variable-size param buckets

* Add dist Adam unit test with contiguous grad buffers

* Optimize compute/communication overlap in dist Adam optim step

* Restore Dist Adam default of splitting params across default-sized buckets

* Support initializing multiple dist Adam param buckets together

The buckets perform communication together, so they are effectively a large bucket.

* Handle recent change in PyTorch API for coalescing NCCL calls

jacobhinkle committed:
Remove explicit cache and Fusion() calls in instance norm

jacobhinkle committed:
Merge remote-tracking branch 'nvidia/master' into py_nvfuser_instance_norm

jacobhinkle committed:
Use dynamic shapes in norm_fusion_{forward,backward}

This allows NVFuser to use a cached Fusion when changing any input
dimension including batch and channel dimensions. The test was updated
to run a few different sizes and check that the cache does not grow.

jacobhinkle committed:
Remove bool entry from torch2datatype

jacobhinkle committed:
Enable running test_instance_norm_nvfuser.py directly

jacobhinkle committed:
Remove unnecessary f-string when setting device

Co-authored-by: Masaki Kozuki <masaki.kozuki.2014@gmail.com>

jacobhinkle committed:
Accept strided inputs to instance norm

This includes some tests for strided inputs. Note that there is still
some 3D-specific code in instance_norm.py. Also note that changing
spatial size currently triggers a new cached Fusion. I will look into
this further before marking this PR ready.

jacobhinkle committed:
Add missing casts for half-precision inputs to instance norm

jacobhinkle committed:
Add missing bias gradient for half-precision instance norm

jacobhinkle committed:
Keep full-precision mean/invstd in instance norm

Also use a simpler method for defining tensors; use size and stride
(this does not actually hard-code concrete sizes).

jacobhinkle committed:
Compare to float64 in instance norm tests

I also simplified a lot of this code and used default tolerances for
assert_close except for half precision.

jacobhinkle committed:
Bump tols for halfs based on 3080 results

jacobhinkle committed:
Use inequality in fc.num_fusions() check

As noted, on multi-gpu systems if the multigpu test is run first then
cache hits will cause num_fusions() to be too low. Both tests now pass
(tests on A40x4). I will now get some benchmarks...

jacobhinkle committed:
Remove check for num_fusions in instance norm test

New fusions may be scheduled when passed different size inputs, so this
test doesn't make sense. Instead, we will pass over a number of
problematic-sized inputs.

jacobhinkle committed:
Force symbolic shapes in instance norm fusion

When passing inputs with singleton dimensions, using sizes and strides
in define_tensor leads to automatically setting broadcast domains for
those dimensions, which causes errors. Instead, we now manually compute
the contiguity for each axis, and use symbolic_sizes/contiguous inputs
for define_tensor.

jacobhinkle committed:
Test in training and inference mode, and more input sizes

jacobhinkle committed:
Convert instance norm tests to use pytest

jacobhinkle committed:
Use stat_dims_nobatch for broadcasting weight in norm backward

This fixes the last remaining bug in the Python NVFuser norm class. All
tests are now passing on Ampere architecture with all dtypes: float16,
float32, float64, bfloat16
